aeron-tools/src/main/java/uk/co/real_logic/aeron/tools/SubscriberTool.java

```diff
@@ -0,0 +1,719 @@
+        {
+            totalMessages += subscribers[i].verifiableMessagesReceived();
+        }
+        return totalMessages;
+    }
+
+    /**
+     * A snapshot (non-atomic) total of the number of bytes
+     * received across all receiving threads.
+     * @return current total number of bytes received
+     */
+    @Override
+    public long bytes()
+    {
+        long totalBytes = 0;
+        for (int i = 0; i < subscribers.length; i++)
+        {
+            totalBytes += subscribers[i].bytesReceived();
+        }
+        return totalBytes;
+    }
+
+    /**
+     * A snapshot (non-atomic) total of the number of non-verifiable messages
+     * received across all receiving threads.
+     * @return current total number of non-verifiable messages received
+     */
+    @Override
+    public long nonVerifiableMessages()
+    {
+        long totalMessages = 0;
+        for (int i = 0; i < subscribers.length; i++)
+        {
+            totalMessages += subscribers[i].nonVerifiableMessagesReceived();
+        }
+        return totalMessages;
+    }
+
+    /**
+     * Optionally sets the random seed used for the TLRandom class, and reports the seed used.
+     * @return the seed to use
+     */
+    @Override
+    public long setSeed(long seed)
+    {
+        if (options.getRandomSeed() != 0)
+        {
+            seed = options.getRandomSeed();
+        }
+        LOG.info("{}", String.format("Thread %s using random seed %d.", Thread.currentThread().getName(), seed));
+        return seed;
+    }
+
+    class SubscriberThread implements Runnable, InactiveConnectionHandler, NewConnectionHandler, RateController.Callback
+    {
+        final int threadId;
+        private long nonVerifiableMessagesReceived;
+        private long verifiableMessagesReceived;
+        private long bytesReceived;
+        private byte[] bytesToWrite = new byte[1];
+        private final Aeron.Context ctx;
+        private final Aeron aeron;
+        private final UnsafeBuffer controlBuffer = new UnsafeBuffer(new byte[4 + 4 + CHANNEL_NAME_MAX_LEN + 4 + 4]);
+        private final RateController rateController;
+
+        /* All subscriptions we're interested in. */
+        final Subscription subscriptions[];
+        /* Just the subscriptions we have reason to believe might have data available - these we actually poll on. */
+        final Subscription activeSubscriptions[];
+        int activeSubscriptionsLength = 0;
+
+        private final Publication controlPublication;
+        private final Subscription controlSubscription;
+        /* Doesn't use TLRandom, since this really does need to be random and shouldn't
+         * be affected by manually setting the seed. */
+        private final int controlSessionId = ThreadLocalRandom.current().nextInt();
+
+        /* channel -> stream ID -> session ID */
+        private final HashMap<String, Int2ObjectHashMap<Int2ObjectHashMap<MessageStream>>> messageStreams =
```

Comment:
Should I use @link even in comments that are normal, non-Javadoc-style code comments like this one?  Will those still get found and changed during a refactor?  Should I convert normal comments to /** javadoc-style comments?  I would expect not, since I don't think we'd want to clutter up the actual generated doc like that.